File: src/services/vertex.ts
import { GoogleGenAI } from "@google/genai";
import { config } from "../config.js";
import type { GenerateResult } from "../types.js";

// Approximate Vertex AI pricing (USD).
// gemini-2.5-flash-lite: $0.10/1M input tokens, $0.40/1M output tokens
// text-embedding-005: $0.000025/1K characters
const PRICE_INPUT_PER_TOKEN = 0.1 / 1_000_000;
const PRICE_OUTPUT_PER_TOKEN = 0.4 / 1_000_000;
const PRICE_EMBED_PER_CHAR = 0.000025 / 1_000;

// text-embedding-005 limits: 250 instances/request, ~30k chars/request
const MAX_TEXTS_PER_BATCH = 250;
const MAX_CHARS_PER_BATCH = 30_000;

let _client: GoogleGenAI | null = null;

function getClient(): GoogleGenAI {
	if (!_client) {
		_client = new GoogleGenAI({
			vertexai: true,
			project: config.googleCloudProject,
			location: config.googleCloudLocation,
		});
	}
	return _client;
}

function makeBatches(texts: string[]): string[][] {
	const batches: string[][] = [];
	let current: string[] = [];
	let currentChars = 0;
	for (const text of texts) {
		const n = text.length;
		if (
			current.length > 0 &&
			(currentChars + n > MAX_CHARS_PER_BATCH ||
				current.length >= MAX_TEXTS_PER_BATCH)
		) {
			batches.push(current);
			current = [text];
			currentChars = n;
		} else {
			current.push(text);
			currentChars += n;
		}
	}
	if (current.length > 0) batches.push(current);
	return batches;
}

export async function embedTexts(
	texts: string[],
	taskType: string = "RETRIEVAL_DOCUMENT",
): Promise<number[][]> {
	const client = getClient();
	const batches = makeBatches(texts);
	const results: number[][] = [];

	for (const batch of batches) {
		const response = await client.models.embedContent({
			model: config.embeddingModel,
			contents: batch,
			config: {
				taskType: taskType as "RETRIEVAL_DOCUMENT" | "RETRIEVAL_QUERY",
			},
		});
		if (!response.embeddings) throw new Error("No embeddings returned");
		for (const emb of response.embeddings) {
			results.push(emb.values ?? []);
		}
	}

	return results;
}

export async function embedQuery(text: string): Promise<number[]> {
	const embeddings = await embedTexts([text], "RETRIEVAL_QUERY");
	return embeddings[0] ?? [];
}

export async function generateText(prompt: string): Promise<GenerateResult> {
	const client = getClient();
	const response = await client.models.generateContent({
		model: config.generationModel,
		contents: prompt,
	});

	const usage = response.usageMetadata;
	const inputTokens = usage?.promptTokenCount ?? 0;
	const outputTokens = usage?.candidatesTokenCount ?? 0;
	const costUsd =
		inputTokens * PRICE_INPUT_PER_TOKEN + outputTokens * PRICE_OUTPUT_PER_TOKEN;

	return {
		text: response.text ?? "",
		inputTokens,
		outputTokens,
		costUsd,
	};
}

export function embedCost(texts: string[]): number {
	return texts.reduce((sum, t) => sum + t.length, 0) * PRICE_EMBED_PER_CHAR;
}

// Cosine similarity between two vectors (no external dep)
export function cosineSimilarity(a: number[], b: number[]): number {
	let dot = 0,
		normA = 0,
		normB = 0;
	for (let i = 0; i < a.length; i++) {
		dot += a[i] * b[i];
		normA += a[i] * a[i];
		normB += b[i] * b[i];
	}
	const denom = Math.sqrt(normA) * Math.sqrt(normB);
	return denom === 0 ? 0 : dot / denom;
}

export function vecNorm(v: number[]): number {
	return Math.sqrt(v.reduce((s, x) => s + x * x, 0));
}

File: hts_classifier/services/vertex.py
import asyncio
import threading
from typing import NamedTuple

from google import genai
from tqdm import tqdm
from google.genai.types import EmbedContentConfig, GenerateContentConfig, ThinkingConfig

from ..core.config import settings

# Thread-local storage: each executor thread gets its own client instance.
# A shared singleton is not safe when run_in_executor fires concurrent threads.
_thread_local = threading.local()

# text-embedding-005 limits: 250 instances per request, 20k tokens per request.
# Short HTS descriptions average ~2 chars/token, so 30k chars ≈ 15k tokens.
_MAX_CHARS_PER_BATCH = 30_000
_MAX_TEXTS_PER_BATCH = 250

# Approximate Vertex AI pricing (USD). Update if model changes.
# gemini-2.5-flash-lite: $0.10/1M input tokens, $0.40/1M output tokens
# text-embedding-005: $0.000025/1K characters
_PRICE_INPUT_PER_TOKEN = 0.10 / 1_000_000
_PRICE_OUTPUT_PER_TOKEN = 0.40 / 1_000_000
_PRICE_EMBED_PER_CHAR = 0.000025 / 1_000


class GenerateResult(NamedTuple):
    text: str
    input_tokens: int
    output_tokens: int

    @property
    def cost_usd(self) -> float:
        return (
            self.input_tokens * _PRICE_INPUT_PER_TOKEN
            + self.output_tokens * _PRICE_OUTPUT_PER_TOKEN
        )


def embed_cost(texts: list[str]) -> float:
    """Approximate embedding cost based on character count."""
    return sum(len(t) for t in texts) * _PRICE_EMBED_PER_CHAR


def get_client() -> genai.Client:
    if not hasattr(_thread_local, "client"):
        _thread_local.client = genai.Client(
            vertexai=True,
            project=settings.google_cloud_project,
            location=settings.google_cloud_location,
        )
    return _thread_local.client


def _embed_batch_sync(texts: list[str], task_type: str) -> list[list[float]]:
    response = get_client().models.embed_content(
        model=settings.embedding_model,
        contents=texts,
        config=EmbedContentConfig(task_type=task_type),
    )
    assert response.embeddings is not None
    return [e.values or [] for e in response.embeddings]


def _generate_sync(prompt: str, model: str | None = None, thinking_level: str | None = None) -> GenerateResult:
    kwargs = {
        "model": model or settings.generation_model,
        "contents": prompt,
    }
    if thinking_level:
        kwargs["config"] = GenerateContentConfig(
            thinking_config=ThinkingConfig(thinking_level=thinking_level)
        )
        
    response = get_client().models.generate_content(**kwargs)
    usage = response.usage_metadata
    return GenerateResult(
        text=response.text or "",
        input_tokens=getattr(usage, "prompt_token_count", 0) or 0,
        output_tokens=getattr(usage, "candidates_token_count", 0) or 0,
    )


def _make_batches(texts: list[str]) -> list[list[str]]:
    """Split texts into batches that stay within the token limit."""
    batches: list[list[str]] = []
    current: list[str] = []
    current_chars = 0
    for text in texts:
        n = len(text)
        if current and (
            current_chars + n > _MAX_CHARS_PER_BATCH
            or len(current) >= _MAX_TEXTS_PER_BATCH
        ):
            batches.append(current)
            current = [text]
            current_chars = n
        else:
            current.append(text)
            current_chars += n
    if current:
        batches.append(current)
    return batches


async def embed_texts(
    texts: list[str],
    task_type: str = "RETRIEVAL_DOCUMENT",
    show_progress: bool = False,
) -> list[list[float]]:
    """Embed texts with dynamic batching, sequential requests."""
    loop = asyncio.get_running_loop()
    batches = _make_batches(texts)
    results: list[list[float]] = []
    it = tqdm(batches, desc="batches") if show_progress else batches
    for batch in it:
        batch_result = await loop.run_in_executor(
            None, _embed_batch_sync, batch, task_type
        )
        results.extend(batch_result)
    return results


async def embed_query(text: str) -> list[float]:
    results = await embed_texts([text], task_type="RETRIEVAL_QUERY")
    return results[0]


async def generate_text(prompt: str, model: str | None = None, thinking_level: str | None = None) -> GenerateResult:
    loop = asyncio.get_running_loop()
    return await loop.run_in_executor(None, _generate_sync, prompt, model, thinking_level)
